### PR TITLE
Bugfix JupyterHub Leader Election

### PR DESCRIPTION
**Pull request by lucferbux**
---
Title: Bugfix JupyterHub Leader Election
---

### Changes proposed in this pull request
As we have implemented High Availability on Jupyterhub, we shifted from 1 to 3 containers with the Leader Election strategy.

With this implementation, we could bump into inconsistencies given that one Pod could be thinking he's still the leader elected, but others could have replaced it due to network problems.

Implement a check to detect if a pod running is still the leader elected, and if not, delete it.

#### Diagram

leader-election

### Issues

* https://issues.redhat.com/browse/ODH-453

### Test instructions

* Check the instructions inside the [testing documentation](https://docs.google.com/document/d/1Ga0khZqr-jGpi5ExPJ0Cz8RawI0INc0UGI6Q2O3fTJc/edit#) about the ODH Manifest deployment
* Use this branch as the base image
* You can use olminstall to deploy to the dev cluster the changes
[TODO: How to check that changes are working]

### CODE REVIEW

**Comment by lucferbux** [comment]
I have found a weir issue, those two images that contain the same s2i code behave different with the leader election strategy:

- quay.io/odh-jupyterhub/jupyterhub-img:v0.3.3 -> Works as intended
- quay.io/modh/odh-jupyterhub-container:v1.1.1-10 -> Doesn't work at all (even though it has the s2i run script that determine the leader election strategy)

**Comment by lucferbux** [comment]
> Is it just me or are the commits for this PR weird?

Yeah I did a last minute change cause I didn't know if it was targeted for the code free, I'm refactoring and updating the PR.

**Comment by lucferbux** [comment]
> @lucferbux Can you rebase to pull in the fix for the JupyterHub CI failures?

Great thanks! Done!

**Comment by lucferbux** [comment]
@LaVLaS I still have issues with the tests. I think the problem lies in selenium, take a look:

```
SessionNotCreatedException: Message: session not created: This version of ChromeDriver only supports Chrome version 91
Current browser version is 93.0.4577.63 with binary path /usr/bin/google-chrome
```

I've bumped into this error in other projects, and it's usually caused by selenium updating the chrome version but the host machine having a prior ChromeDriver. 

I'm not a 100% sure if that's the cause, but if you want I can take a look in that direction!

**Comment by vpavlin** [comment]
> I've bumped into this error in other projects, and it's usually caused by selenium updating the chrome version but the host machine having a prior ChromeDriver.

This is due to some updates in selenium vs. Chrome Driver

Update the `CHROMEDRIVER_VER` in https://github.com/opendatahub-io/odh-manifests/blob/master/tests/Dockerfile to the value in the error message and it should start working again

**Comment by lucferbux** [comment]
> > I've bumped into this error in other projects, and it's usually caused by selenium updating the chrome version but the host machine having a prior ChromeDriver.
> 
> This is due to some updates in selenium vs. Chrome Driver
> 
> Update the `CHROMEDRIVER_VER` in https://github.com/opendatahub-io/odh-manifests/blob/master/tests/Dockerfile to the value in the error message and it should start working again

Great thanks @vpavlin trying now with new version!

**Comment by lucferbux** [comment]
Great, is now working with the fix proposed, i would need aprooval.
@anishasthana @maulikjs @gmfrasca @rimolive

**Comment by vpavlin** [comment]
> Thanks @vpavlin i was thinking about that, this is just the preliminary PR in order to test the issue.

@lucferbux What about my comment that it should not be implemented directly in yaml?:-)

**Comment by lucferbux** [comment]
> > Thanks @vpavlin i was thinking about that, this is just the preliminary PR in order to test the issue.
> 
> @lucferbux What about my comment that it should not be implemented directly in yaml?:-)

Oh yeah, sorry I totally forgot to answer you 🤦‍♂️. OK, first of all, i removed the script in the sidecar container cause it wasn't the right solution, and second, yeah I totally agree, it could be a great solution although there are a few drawbacks.

1. Having script referenced in the yaml make harder to check the functionality of a container at first glance, I'm totally ok with that but it's a thing to consider.
2. Sorry for being a n00b, but I'm not 100% sure how to implement scripts outside yaml rather than copy them in the images and then run them in the location, if there's an easier solution, please let me know @vpavlin 

But yeah, sorry cause I discuss it with someone else but I didn't come back to you, thanks for the reminder!

**Comment by vpavlin** [comment]
No worries:-) 


> 1. Having script referenced in the yaml make harder to check the functionality of a container at first glance, I'm totally ok with that but it's a thing to consider.

I am not sure what you mean by that - the reason for this to be called by livenessProbe is to not have to check "visually", but let the cluster handle the liveness of the pods


> 2. Sorry for being a n00b, but I'm not 100% sure how to implement scripts outside yaml rather than copy them in the images and then run them in the location, if there's an easier solution, please let me know @vpavlin

No other way - only way it to put the script in the image and call it from `livenessProbe`, but it's probably better than having shell code in YAML
> 
> But yeah, sorry cause I discuss it with someone else but I didn't come back to you, thanks for the reminder

**Comment by lucferbux** [comment]
> No worries:-)
> 
> > 1. Having script referenced in the yaml make harder to check the functionality of a container at first glance, I'm totally ok with that but it's a thing to consider.
> 
> I am not sure what you mean by that - the reason for this to be called by livenessProbe is to not have to check "visually", but let the cluster handle the liveness of the pods
> 
> > 1. Sorry for being a n00b, but I'm not 100% sure how to implement scripts outside yaml rather than copy them in the images and then run them in the location, if there's an easier solution, please let me know @vpavlin
> 
> No other way - only way it to put the script in the image and call it from `livenessProbe`, but it's probably better than having shell code in YAML
> 
> > But yeah, sorry cause I discuss it with someone else but I didn't come back to you, thanks for the reminder

Perfect! So I will definitely do that, although, I'm gonna cherry pick the fix testing commit into another PR in order to unblock that path, cause I think @rimolive needs that ASAP.

**Comment by vpavlin** [comment]
Yeah, I think it is ok if we merge this for now if it is necessary, but we should definitely try to get it done "right" ASAP to not introduce techical debt

**Comment by lucferbux** [comment]
> Yeah, I think it is ok if we merge this for now if it is necessary, but we should definitely try to get it done "right" ASAP to not introduce techical debt

Oh no no, no worries, I can work on that and have the changes in this same PR, the "urgent" part was fixing the tests, so I'm working on that right now, thanks again @vpavlin

**Comment by vpavlin** [comment]
> the "urgent" part was fixing the tests,

ah, gotcha

**Comment by lucferbux** [comment]
I've added a 9-second sleep in the liveness probe of the sidecar container, cause I've been testing the changes in my dev cluster and the error triggers in both 3 sidecar containers at the same time, but if their liveness probe are not synced, some of them will be restarted, and hence they will work again and some of them not and they'll keep malfunctioning

**Comment by lucferbux** [comment]
Btw, this PR should not be merged *until* this other PR is approved and there's a new jupyterhub-odh img to link.
https://github.com/opendatahub-io/jupyterhub-odh/pull/116

**Review by vpavlin — CHANGES_REQUESTED** [review]
Once the image is updated, I think we are good to merge

**Review by anishasthana — COMMENTED** [review]
/lgtm

**Review by vpavlin — APPROVED** [review]

**Pull request merged**
